File: Day_01/day1.py
from itertools import combinations
from math import prod

# Solution 1 - prefer this for oneliners
# Part 1
print("Part 1: {}".format([x * y for (x, y) in combinations([int(x) for x in open('data.txt')], 2) if x + y == 2020][0]))
# Part 2
print("Part 2: {}".format([x * y * z for (x, y, z) in combinations([int(x) for x in open('data.txt')], 3) if x + y + z == 2020][0]))


# Solution 2 - creating reusable function
def execute(n):
    print("Part {}: {}".format(n-1, next(prod(x) for x in combinations([int(x) for x in open('data.txt')], n) if sum(x) == 2020)))


# Part 1
execute(2)
# Part 2
execute(3)

File: Day_01/day1.nim
import strutils
import combinations

let file = readFile("data.txt").splitLines()

# Part 1
for x in combinations(file, 2):
    var a = parseInt(x[0])
    var b = parseInt(x[1])
    if a + b == 2020:
        echo "Part 1: ", a * b
        break

# Part 2
for x in combinations(file, 3):
    var a = parseInt(x[0])
    var b = parseInt(x[1])
    var c = parseInt(x[2])
    if a + b + c == 2020:
        echo "Part 2: ", a * b * c
        break
